src/main/java/org/pkcs11/jacknji11/GetAttributeProcess.java

```diff
@@ -0,0 +1,381 @@
+            if (!entry.isFetched()) {
+                query.add(entry.query());
+                if (!batchMode) {
+                    break;
+                }
+            }
+        }
+        return query.toArray(new CKA[0]);
+    }
+
+    /**
+     * Get an entry for given attribute type.
+     *
+     * @param type of attribute
+     * @return entry for given attribute type
+     * @throws IllegalStateException if no such attribute type is present
+     */
+    private Entry getEntry(long type) {
+        for (Entry entry : entries) {
+            if (entry.type == type) {
+                return entry;
+            }
+        }
+        throw new IllegalStateException("No such attribute type: " + type);
+    }
+
+    /**
+     * Main method to fetch attributes.
+     *
+     * @return fetched attributes
+     * @throws CKRException if there is an error fetching attributes
+     */
+    CKA[] fetch() throws CKRException {
+
+        // this is query template that will be sent to the PKCS#11 interface
+        CKA[] query;
+        // loop for as long as there is anything to query
+        while ((query = buildQuery()).length > 0) {
+
+            // fetch attributes using query
+            long rv = cryptoki.GetAttributeValue(session, object, query);
+
+            // if there was an error indicated we need to process it
+            if (rv != CKR.OK) {
+                processError(rv, query);
+            }
+
+            // if we are here it means that there was no actual error,
+            // and we can extract available attributes
+            processAvailable(query);
+
+            // next iteration will either fetch rest of attributes or there is nothing to fetch
+        }
+
+        // return result from fetch process
+        return buildResult();
+    }
+
+    /**
+     * Process error result and response.
+     * <p>
+     * PKCS11 spec says:
+     * <quote>
+     * Note that the error codes CKR_ATTRIBUTE_SENSITIVE, CKR_ATTRIBUTE_TYPE_INVALID, and CKR_BUFFER_TOO_SMALL do not denote true errors for C_GetAttributeValue.
+     * </quote>
+     *
+     * @param rv       result of fetch operation
+     * @param response response from fetch operation
+     * @throws CKRException if there is an actual error fetching attributes
+     */
+    private void processError(long rv, CKA[] response) throws CKRException {
+        if (rv == CKR.BUFFER_TOO_SMALL) {
+            /*
+             This could happen if we are using max length strategy and max lengths are too small.
+             We assume here that all unavailable attributes are in fact reason for buffer too small
+             which depends on the provider implementation but this is how usually providers behave.
+             Even if some of them are actually invalid we will correct that in next iteration.
+            */
+            bufferTooSmall(response);
+        } else if (rv == CKR.ATTRIBUTE_SENSITIVE || rv == CKR.ATTRIBUTE_TYPE_INVALID) {
```

Comment:
In my experience it's not that unusual when working with many different HSMs. Different HSMs handle some attributes differently and you can easily get one of these. But it's handled, it was more about the (very good) java comment :-)

Reply:
I will remove this "general remark" not to trigger any discussions on its rarity.